test/Prototypes/Unicode.swift

```diff
@@ -252,6 +255,351 @@ Base.SubSequence.Iterator.Element == Base.Iterator.Element {
+    OtherCodeUnits.SubSequence.Iterator.Element == OtherCodeUnits.Iterator.Element
+  {
+    self.init(UnicodeStorage(codeUnits, otherEncoding))
+  }
+#endif // false
+
+  // Quick hack to be easier to test
+  init<
+    OtherCU : UnsignedInteger,
+    OtherEncoding : UnicodeEncoding
+  >
+  (
+    codeUnits: [OtherCU], encodedWith otherEncoding: OtherEncoding.Type
+  )
+  where
+    OtherEncoding.EncodedScalar.Iterator.Element == OtherCU
+  {
+    self.init(UnicodeStorage(codeUnits, otherEncoding))
+  }
+
+}
+
+extension SwiftCanonicalString : Unicode {
+  var codeUnits: CodeUnits { return storage.codeUnits }
+
+  typealias ValidUTF8View = Storage.TranscodedView<UTF8>
+  var utf8: ValidUTF8View { return ValidUTF8View(codeUnits) }
+
+  typealias ValidUTF16View = Storage.TranscodedView<UTF16>
+  var utf16: ValidUTF16View { return ValidUTF16View(codeUnits) }
+
+  typealias ExtendedASCII = LazyMapRandomAccessCollection<CodeUnits, UInt32>
+  var extendedASCII: ExtendedASCII {
+    return codeUnits.lazy.map { UInt32($0) }
+  }
+
+  typealias ValidUTF32View = Storage.TranscodedView<UTF32>
+  var utf32: ValidUTF32View { return ValidUTF32View(codeUnits) }
+
+  // typealias EncodedScalars = Storage.EncodedScalars
+  // var encodedScalars: EncodedScalars {
+  //   return storage.scalars
+  // }
+
+  typealias Characters = Storage.Characters
+  var characters: Characters {
+    return storage.characters
+  }
+
+  func isASCII(scan: Bool = true) -> Bool {
+    if isKnownASCII {
+      return true
+    }
+    return scan && false // TODO: perform scan?
+  }
+  func isLatin1(scan: Bool = true) -> Bool {
+    if isKnownLatin1 {
+      return true
+    }
+    return scan && false // TODO: perform scan?
+  }
+  func isNormalizedNFC(scan: Bool = true) -> Bool {
+    // TODO: is this the ideal normal form for us?
+    return true
+  }
+  func isNormalizedNFD(scan: Bool = true) -> Bool {
+    return false
+    // TODO: perform scan perhaps? If every scalar is a whole grapheme, then
+    // this would be true
+  }
+  func isInFastCOrDForm(scan: Bool = true) -> Bool {
+    // FIXME: *almost* all NFC is FCC, but not all ...
+    return true
+  }
+}
+
+// Super dumb comparable conformance...
+extension SwiftCanonicalString : Comparable {
+  static func ==(
+    _ lhs: SwiftCanonicalString, rhs: SwiftCanonicalString
+  ) -> Bool {
+    return lhs.characters.elementsEqual(rhs.characters)
+  }
+  static func <(
+    _ lhs: SwiftCanonicalString, rhs: SwiftCanonicalString
+  ) -> Bool {
+    for (lhsChar, rhsChar) in zip(lhs.characters, rhs.characters) {
+      if lhsChar != rhsChar {
+        return lhsChar < rhsChar
+      }
+    }
+    return lhs.characters.count < rhs.characters.count
+  }
+}
+
+struct String {
+  enum Contents {
+    // Swift canonical string: UTF-16 in TODO-normal-form
+    case canonical(SwiftCanonicalString)
+
+    // 8-bit Latin1
+    case latin1([UInt8]) // TODO: AnyUTF8? UnicodeStorage? Latin1String?
```

Comment:
Design question: wouldn't UTF8 be more useful? I'm sure you guys discussed this already but I feel that UTF16 is almost always the wrong encoding